src/Symfony/Component/AccessToken/Bridge/OAuth/ClientCredentialsProvider.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\AccessToken\Bridge\OAuth;
+
+use Symfony\Component\AccessToken\AccessToken;
+use Symfony\Component\AccessToken\AccessTokenInterface;
+use Symfony\Component\AccessToken\Bridge\AbstractProvider;
+use Symfony\Component\AccessToken\CredentialsInterface;
+use Symfony\Component\AccessToken\Exception\ProviderFetchException;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * OAuth2 "client_credentials" grant connector.
+ *
+ * This provider implements strictly OAuth 2.0 standard, and should work with all
+ * remote services that implement it correctly.
+ *
+ * "client_credentials" workflow may use the "refresh_token" for refreshing tokens
+ * but this is not implemented yet,
+ *
+ * This implementation was tested with:
+ *   - Microsoft Azure OAuth implementation v2.0
+ *
+ * @todo Test with Orange
+ * @todo Test with Google
+ * @todo Test with Salesforce
+ * @todo Test with ... ?
+ *
+ * @author Pierre Rineau <pierre.rineau@processus.org>
+ */
+class ClientCredentialsProvider extends AbstractProvider
+{
+    public function __construct(
+        private readonly HttpClientInterface $httpClient,
+    ) {
+    }
+
+    public function supports(CredentialsInterface $credentials): bool
+    {
+        return $credentials instanceof ClientCredentials;
+    }
+
+    protected function fetchToken(CredentialsInterface $credentials): AccessTokenInterface
+    {
+        \assert($credentials instanceof ClientCredentials);
```

Comment:
i would not use assert here, which depending on php configuration simply is ignored, but do `if (!$credentials instanceof ClientCredentials) {throw new \InvalidArgumentException(...);}`

Reply:
No \InvalidArgumentException, but the one from the component

Reply:
I'm sorry, I often use the `assert()` function only for helping IDE autocompletion: some IDE's don't get the psalm-like type annoations right, asserting make the type explicit. In real life, it's better than a `/** @var Foo $foo */` because unit tests will detect wrong type assertions, and prevent an outdated comment from remaining in code.

If there's no strong convention about this, I'd like to keep those assertions, but I'll comply to the majority.

Reply:
i also like to use assert in some places, usually in places that are somehow not explicit but i can't think of a way how it could happen. here however, i can see the problem happen when a user calls the provider with a wrong token and does not call the supports method first. therefore i would do the instanceof check and throw an actual exception. as oskar notices it should be the exception from this bundle.

Reply:
And I just figured that this asset is actually useless, because the parameter is typed... I must have forgotten it during a refactor.

Reply:
one is the interface, the other the concrete implementation as checked by the supports method. the check is needed i think.